tensorflow/core/kernels/sparse_slice_grad_op_gpu.cu.cc

```diff
@@ -0,0 +1,167 @@
+
+// Helper that wraps a multi-dimensional index and provides a comparison
+// operator that shifts and then compares against another index. This is
+// used for searching for an input index within the output indices.
+struct MultiIndexComparator {
+  EIGEN_DEVICE_FUNC MultiIndexComparator(int rank, const int64_t* indices,
+                                         const int64_t* input_start)
+      : rank_(rank), indices_(indices), input_start_(input_start) {}
+
+  // Note: gpu_helper::lower_bound always compares `haystack < needle`,
+  // so the argument here is always the needle (the input index) and *this
+  // is the haystack (the output index).
+  EIGEN_DEVICE_FUNC bool operator<(const MultiIndexComparator& input) const {
+    for (int d = 0; d < rank_; ++d) {
+      // Shift output index by the slice start.
+      int64_t output_index_i = indices_[d] + input_start_[d];
+      int64_t input_index_i = input.indices_[d];
+      // Lexicographically compare the indexes.
+      if (output_index_i < input_index_i) return true;
+      if (output_index_i > input_index_i) return false;
+    }
+    return false;
+  }
+
+  EIGEN_DEVICE_FUNC bool operator==(const MultiIndexComparator& input) const {
+    for (int d = 0; d < rank_; ++d) {
+      // Shift output index by the slice start.
+      int64_t output_index_i = indices_[d] + input_start_[d];
+      int64_t input_index_i = input.indices_[d];
+      if (output_index_i != input_index_i) return false;
+    }
+    return true;
+  }
+
+ private:
+  int rank_;
+  const int64_t* __restrict__ indices_;
+  const int64_t* __restrict__ input_start_;
+};
+
+struct MultiIndexSearchFunctor {
+  MultiIndexSearchFunctor(int rank, const int64_t* indices,
+                          const int64_t* input_start)
+      : rank_(rank), indices_(indices), input_start_(input_start) {}
+
+  EIGEN_DEVICE_FUNC MultiIndexComparator operator()(int64_t i) const {
+    return {rank_, indices_ + i * rank_, input_start_};
+  }
+
+ private:
+  int rank_;
+  const int64_t* indices_;
+  const int64_t* input_start_;
+};
+
+using IndexIterator = gpuprim::CountingInputIterator<int64_t, int64_t>;
+using MultiIndexSearchIterator =
+    gpuprim::TransformInputIterator<MultiIndexComparator,
+                                    MultiIndexSearchFunctor, IndexIterator>;
+
+template <typename T>
+__global__ void SparseSliceGradKernel(int64_t input_nnz, int64_t output_nnz,
+                                      MultiIndexSearchIterator input_indices,
+                                      MultiIndexSearchIterator output_indices,
+                                      const T* backprop_val_grad, T* val_grad) {
+  for (int64 input_nz : GpuGridRangeX<int64>(input_nnz)) {
+    // Search for the input index in the output indices.
+    // Note: It would be faster to directly test if the input index is within
+    // the slice volume (and also to use flattened indexes), but unfortunately
+    // we don't have the dense shapes so we can't do that.
```

Comment:
But even if you had the dense shapes, you'd still need to call lower_bound to do the binary search, because you need to find the index `output_nz` so you can copy the correct value from `backprop_val_grad` to `val_grad[input_nz]`

Reply:
That's correct; the comment is just pointing out that the test for being inside the slice volume could be done without the need to look up `output_indices[output_nz]`, and that `lower_bound()` would only need to be called inside the true branch.
I've added the word "first" to make this clearer.